blockchain/types/src/avr_history.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Types for reading and writing historical Attestation Verification
+//! Reports from MobileCoin consensus enclaves
+
+use alloc::{string::String, vec::Vec};
+use hex::ToHex;
+use mc_attest_core::VerificationReport;
+use mc_crypto_keys::Ed25519Public;
+use mc_common::HashMap;
+use toml::{map::Map, value::Value};
+use url::Url;
+
+/// Object representing AVR history for all consensus nodes
+pub struct AVRHistoricalRecords {
+    /// AVR Records
+    pub avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>,
+}
+
+impl AVRHistoricalRecords {
+    /// Create new records object
+    pub fn new(avr_records: HashMap<Url, Vec<AVRHistoricalRecord>>) -> Self {
+        Self { avr_records }
+    }
+
+    /// Get a toml representation of the AVR historical records
+    pub fn export_toml_string(&self) -> Result<String, toml::ser::Error>{
+        toml::to_string(&Map::from(self))
+    }
+}
+
+/// Stores a single historical AVR for each tuple of responder_id
+/// & block signing key  in the format specified in draft MCIP #43
+pub struct AVRHistoricalRecord {
+    /// Uri of the consensus node
+    pub node_url: Url,
+
+    /// Block signing key of the consensus enclave
+    pub public_key: Option<Ed25519Public>,
+
+    /// Block the AVR for the signing key becomes valid
+    pub start_block: u64,
+
+    /// Final block the AVR for the signing key is valid
+    pub end_block: u64,
+
+    /// Hex of the AVR bytes
+    pub avr: Option<VerificationReport>,
+}
+
+impl AVRHistoricalRecord {
+    /// Create new AVRHistoricalRecord
+    pub fn new(
+        node_url: &Url,
+        public_key: &Option<Ed25519Public>,
+        start_block: u64,
+        end_block: u64,
+        avr: Option<VerificationReport>,
+    ) -> Self {
+        Self {
+            node_url: node_url.clone(),
+            public_key: *public_key,
+            start_block,
+            end_block,
+            avr,
+        }
+    }
+}
+
+impl Default for AVRHistoricalRecords {
+    fn default() -> Self {
+        let avr_records: HashMap<Url, Vec<AVRHistoricalRecord>> = HashMap::default();
+        Self::new(avr_records)
+    }
+}
+
+// Rust <--> TOML format mapping for all historical AVRs
+impl From<&AVRHistoricalRecords> for Map<String, Value> {
+    fn from(src: &AVRHistoricalRecords) -> Self {
+        let mut toml = Map::new();
+        for (url, records) in src.avr_records.iter() {
+            let maps: Vec<Value> = records.iter().map(|x| Value::Table(Map::from(x))).collect();
+            toml.insert(url.as_str().into(), Value::Array(maps));
```

Comment:
The key should be `avr` or similar constant, not the URL value.

Reply:
The AVRs are collected per node per block range, it makes more sense to work with the responder ID.

Reply:
This is a good point of conversation though. 

An AVR as a key sounds like not a good idea because that definitely doesn't allow any type of ordering that makes any sense because 

1) It'll make both the code and the bootstrap file hard to reason about. For the toml file, we do want people to be able to visually scan it and having that as a table key will be visually illegible. And secondly, from the devs perspective it's also tough to think about what node that corresponds to.

and 

2) It doesn't have much meaningful ordering. While the derive works, it'd order as sig/chain/httpbody which has little meaning order-wise.

Reply:
Have you tried parsing this TOML? I think having a constant key makes parsing/deserializing simpler.

Reply:
Yeah I have! The serialization just leaves the field out of toml and deserializing turns it back into a None if it's not there or Some if it is.